src/StreamSelectLoop.php

```diff
@@ -220,27 +213,112 @@ private function waitForStreamActivity($timeout)
     }
 
     /**
+     * Returns integer amount of seconds in $time.
+     *
+     * @param float $time – time in seconds
+     *
+     * @return int
+     */
+    private static function getSeconds($time)
+    {
+        /*
+         * Workaround for PHP int overflow:
+         * (float)PHP_INT_MAX == PHP_INT_MAX => true
+         * (int)(float)PHP_INT_MAX == PHP_INT_MAX => false
+         * (int)(float)PHP_INT_MAX == PHP_INT_MIN => true
+         */
+        if ($time == PHP_INT_MAX) {
+            return PHP_INT_MAX;
+        }
+
+        return intval(floor($time));
+    }
+
+    /**
+     * Returns integer amount of microseconds in $time.
+     *
+     * @param float $time – time in seconds
+     *
+     * @return int
+     */
+    private static function getMicroseconds($time)
+    {
+        $fractional = fmod($time, 1);
+        $microseconds = round($fractional * self::MICROSECONDS_PER_SECOND);
+
+        return intval($microseconds);
+    }
+
+    /**
+     * Returns integer amount of nanoseconds in $time.
+     * The precision is 1 microsecond.
+     *
+     * @param float $time – time in seconds
+     *
+     * @return int
+     */
+    private static function getNanoseconds($time)
+    {
+        return intval(self::getMicroseconds($time) * self::NANOSECONDS_PER_MICROSECOND);
+    }
+
+    /**
      * Emulate a stream_select() implementation that does not break when passed
      * empty stream arrays.
      *
      * @param array        &$read   An array of read streams to select upon.
      * @param array        &$write  An array of write streams to select upon.
-     * @param integer|null $timeout Activity timeout in microseconds, or null to wait forever.
+     * @param float|null   $timeout Activity timeout in seconds, or null to wait forever.
      *
      * @return integer|false The total number of streams that are ready for read/write.
      * Can return false if stream_select() is interrupted by a signal.
      */
     protected function streamSelect(array &$read, array &$write, $timeout)
     {
+        $seconds = $timeout === null ? null : self::getSeconds($timeout);
+        $microseconds = $timeout === null ? 0 : self::getMicroseconds($timeout);
+        $nanoseconds = $timeout === null ? 0 : self::getNanoseconds($timeout);
```

Comment:
Calculation of `$microseconds` and `$nanoseconds` can probably be moved into the `if`s so they are only calculated when needed.